### PR TITLE
ReplicationPad accepts 0-dim batch size.

### PR DESCRIPTION
**Pull request by v0dro**
This PR patches the ReplicationPad modules in `torch.nn` to be compatible with 0-dim batch sizes.

EDIT: this is part of the work on gh-12013 (make all nn layers accept empty batch size)

### CODE REVIEW

**Comment by v0dro** [comment]
@nikitaved @ngimel @rgommers do have a look.

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 1]
```diff
@@ -216,6 +217,9 @@ Tensor& replication_pad1d_backward_out_cpu_template(
   /* get contiguous gradOutput */
   auto gradOutput = gradOutput_.contiguous();
   gradInput.resize_as_(input);
+  if (gradInput.numel() == 0) {
+    return gradInput;
```
Comment (rgommers) [inline aten/src/ATen/native/ReplicationPadding.cpp:221]: Should this return a clone instead? I think this is fine as is because it's a gradient (which shouldn't be operated on directly inplace), but looking at commit 29fabb1fbc and the discussion in gh-32612 I wanted to comment for another reviewer to make sure.

File: torch/testing/_internal/common_nn.py [hunk 2]
```diff
@@ -1961,12 +1961,33 @@ def fractional_max_pool3d_test(test_case):
         cpp_constructor_args='torch::nn::ReplicationPad1dOptions({1, 2})',
         input_size=(2, 3, 4),
     ),
```
[the remark this answers is not in the document]
Comment (rgommers) [reply torch/testing/_internal/common_nn.py:1963]: This is also true for the existing tests by the way (`BatchNorm*`, `Conv*`), try `pytest test/test_nn.py -k zero_batch -v`. Looks like that needs fixing.

**Review by rgommers — COMMENTED** [review]
Thanks @v0dro, overall looks good. I think there's a test coverage issue.

File: torch/testing/_internal/common_nn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (v0dro) [reply torch/testing/_internal/common_nn.py:1963]: OK will do. I'll add specific testing using `_test_module_empty_input`. I think the CUDA skipping should be resolved as a separate issue.

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 3]
```diff
@@ -557,8 +568,11 @@ static inline void shapeCheck3d(
   int dimd = 1;
   int dimslices = 0;
 
-  TORCH_CHECK(input.numel() > 0 && (input.dim() == 4 || input.dim() == 5),
-      "non-empty 4D or 5D (batch mode) tensor expected for input, but got: ", input);
+  // allow batch size of 0-dim.
+  TORCH_CHECK(
+      (input.dim() == 4 && input.size(1) != 0 && input.size(2) != 0 && input.size(3) != 0) ||
```
Comment (ngimel) [inline aten/src/ATen/native/ReplicationPadding.cpp:573]: nit: maybe use std::accumulate with std::multiplies here and in similar places, so that you don't have to specialize for dimensions.

File: test/test_nn.py [hunk 4]
```diff
@@ -9382,6 +9382,18 @@ def test_GroupNorm_empty(self, device):
             with torch.backends.cudnn.flags(enabled=False):
                 self._test_module_empty_input(mod, inp)
 
+    def test_ReplicationPad_empty(self, device):
+        for mod, inp in [
+                (torch.nn.ReplicationPad1d(3), torch.randn(0, 3, 10)),
+                (torch.nn.ReplicationPad2d(3), torch.randn(0, 3, 10, 10)),
+                (torch.nn.ReplicationPad3d(3), torch.randn(0, 3, 10, 10, 10))]:
+            mod = mod.to(device)
```
Comment (ngimel) [inline test/test_nn.py:9390]: nit: replication pad does not have parameters, so you don't have to move it to device

File: test/test_nn.py [hunk 5]
```diff
@@ -9382,6 +9382,18 @@ def test_GroupNorm_empty(self, device):
...
+            inp = inp.to(device)
+            self._test_module_empty_input(mod, inp, check_size=False)
+            if self.device_type == 'cuda' and self.has_cudnn():
```
Comment (ngimel) [inline test/test_nn.py:9393]: replicationPad never uses cudnn so you don't have to run tests with cudnn enabled/disabled

File: torch/testing/_internal/common_nn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/testing/_internal/common_nn.py:1963]: @rgommers I've just run the tests and cuda tests are not skipped on my build - perhaps your build does not have cuda enabled?

[the remark this answers is not in the document]
Comment (ngimel) [reply torch/testing/_internal/common_nn.py:1963]: Adding empty batch tests both here and above with _test_module_empty_input is an overkill, you can pick one place. NewModuleTests are more comprehensive, but I'm not sure that you need that.

File: aten/src/ATen/native/ReplicationPadding.cpp (hunk 1 above)
Comment (rgommers) [reply aten/src/ATen/native/ReplicationPadding.cpp:221]: Thanks for confirming.

File: torch/testing/_internal/common_nn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply torch/testing/_internal/common_nn.py:1963]: > perhaps your build does not have cuda enabled?

oops, you're right - I failed to check that. sorry for the noise

File: aten/src/ATen/native/ReplicationPadding.cpp (hunk 3 above)
Comment (v0dro) [reply aten/src/ATen/native/ReplicationPadding.cpp:573]: I think `all_of` makes the code less terse since it only accepts a starting and ending iterator. However I think replacing the long conditions with a single bool variable can be a good idea.

**Comment by v0dro** [comment]
@ngimel could you please review this one again?

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 6]
```diff
@@ -71,9 +71,12 @@ void replication_pad1d_out_cpu_template(
   int pad_l = paddingSize[0];
   int pad_r = paddingSize[1];
 
-  TORCH_CHECK(input_.numel() > 0
-      && (input_.ndimension() == 2 || input_.ndimension() == 3),
-      "non-empty 2D or 3D (batch mode) tensor expected for input");
+  
+  // allow empty batch size but not other dimensions.
+  TORCH_CHECK(
```
Comment (ngimel) [inline aten/src/ATen/native/ReplicationPadding.cpp:76]: If we want to allow only empty batch but not other dimensions, the check here is wrong, should be 
```
auto size_i = input_.size();
TORCH_CHECK((size_i.size()==2 || size_i.size()==3) &&  
std::all_of(size_i.rbegin(), size_i.rbegin()+2, [](auto i){return i!=0;}));
```
This covers both batch mode and non-batch mode. In non-batch mode (2D) you want all values to be non-zero, in the batch mode (3D) you allow only the 0-th element to potentially be 0. The error message also should be clearer - you are erroring out if some of the dimensions are 0, but the error message does not reflect that. 
Same goes for other checks - they all allow more zeros than is right, and all don't provide clear error message.

**Review by ngimel — COMMENTED** [review]
Checks are still not exactly right, otherwise looks good

File: aten/src/ATen/native/ReplicationPadding.cpp (hunk 6 above)
Comment (v0dro) [reply aten/src/ATen/native/ReplicationPadding.cpp:76]: OK this one is wrong and I have corrected it . However, the rest of the checks seem to make sense to me.

* https://github.com/pytorch/pytorch/pull/39137/files#diff-3e39c997ceefe5db9d4b51504dc8c91cR574 checks for dims 1,2,3 for 4D tensor and 1,2,3,4 for 5D tensor.
* This one too: https://github.com/pytorch/pytorch/pull/39137/files#diff-01f6085523cbc69ca71060d98e22f257R364

**Comment by v0dro** [comment]
@ngimel could you please review this again?

**Comment by v0dro** [comment]
@ngimel @rgommers ping!

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 7]
```diff
@@ -71,9 +71,10 @@ void replication_pad1d_out_cpu_template(
   int pad_l = paddingSize[0];
   int pad_r = paddingSize[1];
 
-  TORCH_CHECK(input_.numel() > 0
-      && (input_.ndimension() == 2 || input_.ndimension() == 3),
-      "non-empty 2D or 3D (batch mode) tensor expected for input");
+  // allow empty batch size but not other dimensions.
+  TORCH_CHECK((input_.dim() == 2) ||
+              (input_.dim() == 3 && input_.size(1) != 0 && input_.size(2) != 0),
+              "2D or 3D (batch mode) tensor for input, but got: ", input_.sizes());
```
Comment (ngimel) [inline aten/src/ATen/native/ReplicationPadding.cpp:77]: You should check dimensions 0,1 here for a 2d tensor, and 1,2 for a 3d tensor, please see my comment from previous review.

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 8]
```diff
@@ -339,7 +342,11 @@ void replication_pad2d_out_cpu_template(Tensor& output,
   int dimslices = 0;
   int64_t nbatch = 1;
 
-  TORCH_CHECK(input_.numel() > 0 && (input_.dim() == 3 || input_.dim() == 4),
+  // allow 0 dim batch size and nothing else.
+  bool valid_dims = input_.size(1) != 0 && input_.size(2) != 0;
```
Comment (ngimel) [inline aten/src/ATen/native/ReplicationPadding.cpp:346]: for 3d tensor dims 0,1,2 should be non-zero, for 4d tensor dims 1,2,3 should be nonzero

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 9]
```diff
@@ -557,8 +568,12 @@ static inline void shapeCheck3d(
   int dimd = 1;
   int dimslices = 0;
 
-  TORCH_CHECK(input.numel() > 0 && (input.dim() == 4 || input.dim() == 5),
-      "non-empty 4D or 5D (batch mode) tensor expected for input, but got: ", input);
+  // allow batch size of 0-dim.
+  bool valid_dims = input.size(1) != 0 && input.size(2) != 0 && input.size(3) != 0;
```
Comment (ngimel) [inline aten/src/ATen/native/ReplicationPadding.cpp:572]: for 4d tensor dimensions 0,1,2,3 should be nonzero (there's no batch dimension, so tensor should be non-empty), for 5d tensor dimensions 1,2,3,4 should be nonzero.

**Comment by v0dro** [comment]
I'm sorry the difference in the usage of dimensions for different ReplicationPad classes was a little confusing. I have corrected it now.

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 10]
```diff
@@ -71,9 +71,10 @@ void replication_pad1d_out_cpu_template(
   int pad_l = paddingSize[0];
   int pad_r = paddingSize[1];
 
-  TORCH_CHECK(input_.numel() > 0
-      && (input_.ndimension() == 2 || input_.ndimension() == 3),
-      "non-empty 2D or 3D (batch mode) tensor expected for input");
+  // allow empty batch size but not other dimensions.
+  TORCH_CHECK((input_.dim() == 2 && input_.size(0) != 0 && input_.size(1) != 0) ||
+              (input_.dim() == 3 && input_.size(1) != 0 && input_.size(2) != 0),
+              "2D or 3D (batch mode) tensor for input, but got: ", input_.sizes());
```
Comment (ngimel) [inline aten/src/ATen/native/ReplicationPadding.cpp:77]: Can you please make error message more descriptive? "Expected 2D or 3D (batch mode) tensor for input with possibly 0 batch dimension and other dimensions non-zero, but got" 
Same for other error messages. Also, you seem to test 2D tensor and expect NotImplementedError there, so is this error message even true?

**Review by ngimel — COMMENTED** [review]
Thank you, the dimension checks are correct now, can you please make error messages a little bit clearer?

File: aten/src/ATen/native/ReplicationPadding.cpp (hunk 10 above)
Comment (v0dro) [reply aten/src/ATen/native/ReplicationPadding.cpp:77]: The replication pad 1D function implementation in C++ and CUDA has functionality for 2D tensors but seems to be disabled from the `torch.nn` Python interface.

Comment (v0dro) [reply aten/src/ATen/native/ReplicationPadding.cpp:77]: OK I'll change the error messages.

**Comment by v0dro** [comment]
@ngimel request re-review.

**Review by ngimel — APPROVED** [review]
